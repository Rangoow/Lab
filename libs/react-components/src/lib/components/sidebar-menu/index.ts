export * from './sidebar-menu.component';
export * from './interfaces';
export * from './sidebar-menu-button';
