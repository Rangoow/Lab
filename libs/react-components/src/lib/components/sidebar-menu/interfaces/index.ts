export * from './sidebar-menu-element.interface';
