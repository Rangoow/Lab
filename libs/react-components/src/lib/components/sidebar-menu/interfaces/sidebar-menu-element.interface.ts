import { VitamixId } from '@vtmn/icons/dist/vitamix/font/vitamix';

export interface ISidebarMenuElement {
  readonly key: string;
  readonly text: string;
  readonly title: string;
  readonly icon: VitamixId;
  readonly redirectPath: string;
}
