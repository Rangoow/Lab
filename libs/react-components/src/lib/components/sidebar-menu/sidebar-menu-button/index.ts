export * from './sidebar-menu-button.component';
