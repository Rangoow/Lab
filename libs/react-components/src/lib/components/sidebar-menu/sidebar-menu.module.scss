@import '../../styles/breakpoints';

@media (max-width: mediaquery(mobile-max)) {
  .hiddenMobile {
    display: none;
  }
}

.menu {
  background: var(--vtmn-semantic-color_background-primary);
  width: 100%;
  height: 100%;

  @media (min-width: mediaquery(tablet-min)) {
    flex-direction: column;
  }

  @media (max-width: mediaquery(mobile-max)) {
    flex-direction: row;
  }

  &__list {
    width: 100%;
    height: 100%;

    @media (min-width: mediaquery(tablet-min)) {
      flex-direction: column;
      overflow-y: auto;
    }

    @media (max-width: mediaquery(mobile-max)) {
      flex-direction: row;
    }
  }
}

:global(.vtmn-btn).toggle-button {
  border-radius: 0;
}
