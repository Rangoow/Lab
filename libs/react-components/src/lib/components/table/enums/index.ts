export * from './demo-boolean-tag.enums';
export * from './demo-status-tag.enums';
