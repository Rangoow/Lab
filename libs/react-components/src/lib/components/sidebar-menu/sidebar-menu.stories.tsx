import { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import { SidebarMenu } from './sidebar-menu.component';

const meta: Meta<typeof SidebarMenu> = {
  title: 'Structure/Sidebar Menu',
  component: SidebarMenu,
  decorators: [
    // eslint-disable-next-line @typescript-eslint/naming-convention
    Story => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
};

type Story = StoryObj<typeof SidebarMenu>;

export default meta;

export const Simple: Story = {
  args: {
    menuButtonsProperties: [
      {
        key: 'sidebar-menu_dashboard',
        icon: 'home-line',
        text: 'page.requests-dashboard.title',
        title: 'sidebar-menu.requests-dashboard-button',
        redirectPath: 'requests-dashboard',
      },
      {
        key: 'sidebar-menu_admin',
        icon: 'lock-line',
        text: 'page.admin.title',
        title: 'sidebar-menu.admin-button',
        redirectPath: 'admin',
      },
    ],
  },
};
