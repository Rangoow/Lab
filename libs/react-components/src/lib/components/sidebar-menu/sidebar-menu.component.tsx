import { VtmnButton, VtmnIcon } from '@vtmn/react';
import classNames from 'classnames/bind';
import { ReactElement, useCallback, useMemo, useState } from 'react';
import { EReactComponentsDataTestId } from '../../enums/';
import { ISidebarMenuElement } from './interfaces';
import { SidebarMenuButton } from './sidebar-menu-button';
import styles from './sidebar-menu.module.scss';

interface IProps {
  readonly menuButtonsProperties: ISidebarMenuElement[];
  readonly menuButtonTitle: string;
}

export function SidebarMenu({ menuButtonsProperties, menuButtonTitle }: IProps) {
  const [isMenuCollapsed, setIsMenuCollapsed] = useState<boolean>(true);
  const toggleMenu = useCallback((): void => {
    setIsMenuCollapsed(!isMenuCollapsed);
  }, [isMenuCollapsed]);
  const cx = classNames.bind(styles);

  return useMemo(
    () => (
      <div role="menubar" className={cx('menu', 'flex')}>
        <div className={cx('hiddenMobile')}>
          <VtmnButton
            variant="ghost"
            onClick={toggleMenu}
            aria-label="sidebar_menu_button"
            title={menuButtonTitle}
            size="stretched"
            className={cx('toggle-button')}
          >
            <VtmnIcon
              data-testid={EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON}
              value={isMenuCollapsed ? `menu-line` : `chevron-left-line`}
            />
          </VtmnButton>
        </div>
        <ul className={cx('menu__list', 'flex')}>
          {menuButtonsProperties.map(
            (element: ISidebarMenuElement): ReactElement => (
              <li key={element.key}>
                <SidebarMenuButton
                  title={element.title}
                  icon={element.icon}
                  text={element.text}
                  redirectPath={element.redirectPath}
                  isCollapsed={isMenuCollapsed}
                />
              </li>
            ),
          )}
        </ul>
      </div>
    ),
    [cx, isMenuCollapsed, menuButtonTitle, menuButtonsProperties, toggleMenu],
  );
}
