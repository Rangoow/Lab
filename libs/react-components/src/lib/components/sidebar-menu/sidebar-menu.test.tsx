import { randDirectoryPath, randSentence, randText, randUuid, randWord } from '@ngneat/falso';
import { RenderResult, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { EReactComponentsDataTestId } from '../../enums';
import { customRender } from '../../tests/custom-render';
import { SidebarMenu } from './sidebar-menu.component';
import { ISidebarMenuElement } from './interfaces';

const user = userEvent.setup();

describe(`uiSidebarMenu`, (): void => {
  let menuButtons: ISidebarMenuElement[];
  let menuButtonTitle: string;

  describe('snapshot testing', (): void => {
    beforeEach((): void => {
      menuButtons = [
        {
          key: `sidebar-menu_dashboard`,
          icon: `home-line`,
          text: `Dashboard`,
          title: `Go to dashboard`,
          redirectPath: `/`,
        },
        {
          key: `sidebar-menu_settings`,
          icon: `settings-line`,
          text: `Settings`,
          title: `Go to settings`,
          redirectPath: `/settings`,
        },
      ];
      menuButtonTitle = 'menu button';
    });

    it(`should render correctly and match the snapshot`, async (): Promise<void> => {
      expect.assertions(1);

      const result: RenderResult = await customRender(
        <SidebarMenu menuButtonTitle={menuButtonTitle} menuButtonsProperties={menuButtons} />,
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('component testing', (): void => {
    const menuButton1Text: string = randWord();
    const menuButton2Text: string = randWord();
    const menuButton1Title: string = randSentence();
    const menuButton2Title: string = randSentence();

    beforeEach((): void => {
      menuButtons = [
        {
          key: randUuid(),
          icon: `home-line`,
          text: menuButton1Text,
          title: menuButton1Title,
          redirectPath: randDirectoryPath(),
        },
        {
          key: randUuid(),
          icon: `settings-line`,
          text: menuButton2Text,
          title: menuButton2Title,
          redirectPath: randDirectoryPath(),
        },
      ];
      menuButtonTitle = randText();
    });

    it(`should render all the given menu buttons`, async (): Promise<void> => {
      expect.assertions(4);

      await customRender(<SidebarMenu menuButtonTitle={menuButtonTitle} menuButtonsProperties={menuButtons} />);

      const menuButton: HTMLElement = screen.getByRole(`button`, { name: 'sidebar_menu_button' });
      const menuButtonIcon: HTMLElement = screen.getByTestId(EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON);
      const button1: HTMLElement = screen.getByRole(`button`, { name: menuButton1Title });
      const button2: HTMLElement = screen.getByRole(`button`, { name: menuButton2Title });

      expect(menuButton).toBeVisible();
      expect(button1).toBeVisible();
      expect(button2).toBeVisible();
      expect(menuButtonIcon).toHaveClass(`vtmx-menu-line`);
    });

    describe(`when the user click on the menu button`, (): void => {
      it(`should toggle the display and change the menu button icon`, async (): Promise<void> => {
        expect.assertions(3);

        await customRender(<SidebarMenu menuButtonTitle={menuButtonTitle} menuButtonsProperties={menuButtons} />);
        const menuButton: HTMLElement = screen.getByRole(`button`, { name: 'sidebar_menu_button' });
        const menuButtonIcon: HTMLElement = screen.getByTestId(EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON);

        await user.click(menuButton);

        const button1: HTMLElement = screen.getByText(menuButton1Text);
        const button2: HTMLElement = screen.getByText(menuButton2Text);

        expect(menuButtonIcon).toHaveClass(`vtmx-chevron-left-line`);
        expect(button1).toBeVisible();
        expect(button2).toBeVisible();
      });
    });

    describe(`when the user click on the menu button two times`, (): void => {
      it(`should toggle the display and change the menu button icon`, async (): Promise<void> => {
        expect.assertions(1);

        await customRender(<SidebarMenu menuButtonTitle={menuButtonTitle} menuButtonsProperties={menuButtons} />);
        const menuButton: HTMLElement = screen.getByRole(`button`, { name: 'sidebar_menu_button' });
        const menuButtonIcon: HTMLElement = screen.getByTestId(EReactComponentsDataTestId.SIDEBAR_MENU_BUTTON);

        await user.click(menuButton);
        await user.click(menuButton);

        expect(menuButtonIcon).toHaveClass(`vtmx-menu-line`);
      });
    });
  });
});
