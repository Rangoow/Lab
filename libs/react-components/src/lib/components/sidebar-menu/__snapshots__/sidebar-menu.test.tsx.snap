// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`uiSidebarMenu > snapshot testing > should render correctly and match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="menu flex"
        role="menubar"
      >
        <div
          class="hiddenMobile"
        >
          <button
            aria-label="sidebar_menu_button"
            class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched toggle-button   "
            title="menu button"
          >
            <span
              class="vtmx-menu-line "
              data-testid="react--components--sidebar--menu-button"
              style="font-size: 24px;"
            />
          </button>
        </div>
        <ul
          class="menu__list flex"
        >
          <li>
            <button
              class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
              title="Go to dashboard"
            >
              <span
                class="vtmx-home-line "
                data-testid="react--components--sidebar--menu--button--icon"
                style="font-size: 24px;"
              />
            </button>
          </li>
          <li>
            <button
              class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
              title="Go to settings"
            >
              <span
                class="vtmx-settings-line "
                data-testid="react--components--sidebar--menu--button--icon"
                style="font-size: 24px;"
              />
            </button>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="menu flex"
      role="menubar"
    >
      <div
        class="hiddenMobile"
      >
        <button
          aria-label="sidebar_menu_button"
          class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched toggle-button   "
          title="menu button"
        >
          <span
            class="vtmx-menu-line "
            data-testid="react--components--sidebar--menu-button"
            style="font-size: 24px;"
          />
        </button>
      </div>
      <ul
        class="menu__list flex"
      >
        <li>
          <button
            class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
            title="Go to dashboard"
          >
            <span
              class="vtmx-home-line "
              data-testid="react--components--sidebar--menu--button--icon"
              style="font-size: 24px;"
            />
          </button>
        </li>
        <li>
          <button
            class="vtmn-btn vtmn-btn_variant--ghost vtmn-btn_size--stretched button--collapsed   "
            title="Go to settings"
          >
            <span
              class="vtmx-settings-line "
              data-testid="react--components--sidebar--menu--button--icon"
              style="font-size: 24px;"
            />
          </button>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
